turbine/src/addr_cache.rs

```diff
@@ -0,0 +1,412 @@
+            ShredId::new(slot, index as u32, shred_type)
+        }
+        if self.counts.len() == 1 {
+            let slot = self.counts.keys().next().copied().unwrap();
+            return self
+                .get_cache_entry_mut(slot)
+                .get_shreds(EXTEND_BUFFER)
+                .take(num_shreds)
+                .map(|entry| make_shred(slot, entry))
+                .collect();
+        }
+        let mut counts: Vec<(/*count:*/ usize, Slot)> = self
+            .counts
+            .iter()
+            .map(|(&slot, &count)| (count, slot))
+            .collect();
+        counts.sort_unstable();
+        let mut out = Vec::with_capacity(num_shreds);
+        while let Some(count) = num_shreds.checked_sub(out.len()).filter(|&k| k > 0) {
+            let Some((_, slot)) = counts.pop() else {
+                break;
+            };
+            // Leave some capacity for the 2nd most frequent slot.
+            let count = count.min(num_shreds * 3 / 4);
+            out.extend(
+                self.get_cache_entry_mut(slot)
+                    .get_shreds(EXTEND_BUFFER)
+                    .take(count)
+                    .map(|entry| make_shred(slot, entry)),
+            );
+        }
+        out
+    }
+
+    // Returns a mutable reference to the cached entry for the given slot.
+    // Initializes the entry if not allocated yet.
+    #[inline]
+    fn get_cache_entry_mut(&mut self, slot: Slot) -> &mut CacheEntry {
+        self.cache
+            .entry(slot)
+            .or_insert_with(|| CacheEntry::new(ADDR_CAPACITY))
+    }
+
+    // If there are more than ROLLING_WINDOW_NUM_SHREDS shreds in the rolling
+    // window, drops the oldest entries and updates self.counts.
+    fn maybe_trim_slot_counts(&mut self) {
+        while let Some(count) = self
+            .num_shreds
+            .checked_sub(ROLLING_WINDOW_NUM_SHREDS)
+            .filter(|&k| k > 0)
+        {
+            let (slot, num_shreds) = self.window.front_mut().unwrap();
+            let count = count.min(*num_shreds);
+            self.num_shreds -= count;
+            *num_shreds -= count;
+            let Entry::Occupied(mut entry) = self.counts.entry(*slot) else {
+                panic!("Entry must exist if it has non-zero count.");
+            };
+            *entry.get_mut() -= count;
+            if *entry.get() == 0 {
+                entry.remove_entry();
+            }
+            if *num_shreds == 0 {
+                self.window.pop_front();
+            }
+        }
+    }
+
+    // If there are more than 2 * self.capacity entries in the cache, drops
+    // the slots with the fewest counts in the rolling window.
+    fn maybe_trim_cache(&mut self) {
+        if self.cache.len() <= self.capacity.saturating_mul(2) {
+            return;
+        }
+        let mut entries: Vec<((Slot, CacheEntry), /*count:*/ usize)> = self
+            .cache
+            .drain()
```

Comment:
might be better to avoid the `drain` here by collecting just the `(count, slot)` to find the cutoff, and `retain` instead of `extend` below

Reply:
But then we need to allocate another hash-map to look-up which slots we want to retain, no?

Also, this cache is pretty small, like ~9 slots currently. So whatever we do it won't have a noticeable impact anyways.